ios/TwilioVoiceReactNative.m

```diff
@@ -439,23 +444,60 @@ - (void)sendEventWithName:(NSString *)eventName body:(id)body {
         self.deviceTokenData = [testDeviceToken dataUsingEncoding:NSUTF8StringEncoding];
     }
 #endif
+    
+    if (self.registrationInProgress) {
+        reject(kTwilioVoiceReactNativeVoiceError, @"Registration in progress. Please try again later", nil);
+        return;
+    }
 
-    [TwilioVoiceSDK registerWithAccessToken:accessToken
-                                deviceToken:self.deviceTokenData
-                                 completion:^(NSError *error) {
-        if (error) {
-            NSString *errorMessage = [NSString stringWithFormat:@"Failed to register: %@", error];
-            NSLog(@"%@", errorMessage);
-            [self sendEventWithName:kTwilioVoiceReactNativeScopeVoice
-                               body:@{kTwilioVoiceReactNativeVoiceEventType: kTwilioVoiceReactNativeVoiceEventError,
-                                      kTwilioVoiceReactNativeVoiceErrorKeyError: @{kTwilioVoiceReactNativeVoiceErrorKeyCode: @(error.code),
-                                                                                   kTwilioVoiceReactNativeVoiceErrorKeyMessage: [error localizedDescription]}}];
+    self.registrationInProgress = YES;
+    
+    [self asyncPushRegistryInitialization:kPushRegistryDeviceTokenRetryTimeout
+                               completion:^(NSData *deviceTokenData) {
+        if (deviceTokenData) {
+            [TwilioVoiceSDK registerWithAccessToken:accessToken
+                                        deviceToken:deviceTokenData
+                                         completion:^(NSError *error) {
+                if (error) {
+                    NSString *errorMessage = [NSString stringWithFormat:@"Failed to register: %@", error];
+                    NSLog(@"%@", errorMessage);
+                    [self sendEventWithName:kTwilioVoiceReactNativeScopeVoice
+                                       body:@{kTwilioVoiceReactNativeVoiceEventType: kTwilioVoiceReactNativeVoiceEventError,
+                                              kTwilioVoiceReactNativeVoiceErrorKeyError: @{kTwilioVoiceReactNativeVoiceErrorKeyCode: @(error.code),
+                                                                                           kTwilioVoiceReactNativeVoiceErrorKeyMessage: [error localizedDescription]}}];
+                    reject(kTwilioVoiceReactNativeVoiceError, errorMessage, nil);
+                } else {
+                    resolve(nil);
+                }
+                self.registrationInProgress = NO;
```

Comment:
Could we do this before resolving or rejecting? Will it cause race conditions otherwise?